File: src/volatility_filter/models/modules.py
"""Pydantic models for system modules."""

from datetime import datetime
from typing import List, Optional, Dict, Any, Literal
from pydantic import BaseModel, Field

ModuleType = Literal["Data Source", "Function", "Risk", "Strategy", "Execution", "Datasource"]
ModuleStatus = Literal["Active", "Inactive"]


class ModuleConfig(BaseModel):
    """Configuration for a module."""
    parameters: Dict[str, Any] = Field(default_factory=dict)
    credentials: Optional[Dict[str, str]] = None
    settings: Dict[str, Any] = Field(default_factory=dict)


class Module(BaseModel):
    """System module model."""
    id: str
    type: ModuleType
    name: str
    description: str
    version: str
    status: ModuleStatus
    tags: List[str]
    icon: Optional[str] = None
    config: Optional[ModuleConfig] = None
    created_at: datetime = Field(default_factory=datetime.now)
    updated_at: datetime = Field(default_factory=datetime.now)
    
    class Config:
        json_encoders = {
            datetime: lambda v: v.isoformat()
        }


class ModuleStats(BaseModel):
    """Statistics for a module."""
    module_id: str
    execution_count: int = 0
    success_count: int = 0
    failure_count: int = 0
    last_execution: Optional[datetime] = None
    average_execution_time: Optional[float] = None
    last_error: Optional[str] = None


class ModuleExecution(BaseModel):
    """Module execution record."""
    id: str
    module_id: str
    started_at: datetime
    completed_at: Optional[datetime] = None
    status: Literal["running", "success", "failed"] = "running"
    error: Optional[str] = None
    result: Optional[Dict[str, Any]] = None
    duration_ms: Optional[int] = None


class ModuleCreateRequest(BaseModel):
    """Request to create a new module."""
    type: ModuleType
    name: str
    description: str
    version: str = "1.0"
    tags: List[str] = Field(default_factory=list)
    icon: Optional[str] = None
    config: Optional[ModuleConfig] = None


class ModuleUpdateRequest(BaseModel):
    """Request to update a module."""
    name: Optional[str] = None
    description: Optional[str] = None
    version: Optional[str] = None
    status: Optional[ModuleStatus] = None
    tags: Optional[List[str]] = None
    icon: Optional[str] = None
    config: Optional[ModuleConfig] = None


class ModuleResponse(BaseModel):
    """Response containing module data."""
    module: Module
    stats: Optional[ModuleStats] = None


class ModulesListResponse(BaseModel):
    """Response containing list of modules."""
    modules: List[Module]
    total: int
    page: int = 1
    per_page: int = 50

File: volatility-web/lib/api/modules.ts
import { API_BASE_URL } from '../config'

export interface Module {
  id: string
  type: 'Data Source' | 'Function' | 'Risk' | 'Strategy' | 'Execution' | 'Datasource'
  name: string
  description: string
  version: string
  status: 'Active' | 'Inactive'
  tags: string[]
  icon?: string
  created_at?: string
  updated_at?: string
}

export interface ModuleStats {
  module_id: string
  execution_count: number
  success_count: number
  failure_count: number
  last_execution?: string
  average_execution_time?: number
  last_error?: string
}

export interface ModuleResponse {
  module: Module
  stats?: ModuleStats
}

export interface ModulesListResponse {
  modules: Module[]
  total: number
  page: number
  per_page: number
}

export interface ModuleCreateRequest {
  type: Module['type']
  name: string
  description: string
  version?: string
  tags?: string[]
  icon?: string
}

export interface ModuleUpdateRequest {
  name?: string
  description?: string
  version?: string
  status?: Module['status']
  tags?: string[]
  icon?: string
}

class ModulesAPI {
  private baseUrl = `${API_BASE_URL}/api/system-modules`

  async getModules(params?: {
    module_type?: Module['type']
    status?: Module['status']
    search?: string
    page?: number
    per_page?: number
  }): Promise<ModulesListResponse> {
    const searchParams = new URLSearchParams()
    
    if (params?.module_type) searchParams.append('module_type', params.module_type)
    if (params?.status) searchParams.append('status', params.status)
    if (params?.search) searchParams.append('search', params.search)
    if (params?.page) searchParams.append('page', params.page.toString())
    if (params?.per_page) searchParams.append('per_page', params.per_page.toString())
    
    const response = await fetch(`${this.baseUrl}?${searchParams}`)
    if (!response.ok) {
      throw new Error(`Failed to fetch modules: ${response.statusText}`)
    }
    return response.json()
  }

  async getModule(id: string): Promise<ModuleResponse> {
    const response = await fetch(`${this.baseUrl}/${id}`)
    if (!response.ok) {
      throw new Error(`Failed to fetch module: ${response.statusText}`)
    }
    return response.json()
  }

  async createModule(data: ModuleCreateRequest): Promise<ModuleResponse> {
    const response = await fetch(this.baseUrl, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(data),
    })
    if (!response.ok) {
      throw new Error(`Failed to create module: ${response.statusText}`)
    }
    return response.json()
  }

  async updateModule(id: string, data: ModuleUpdateRequest): Promise<ModuleResponse> {
    const response = await fetch(`${this.baseUrl}/${id}`, {
      method: 'PUT',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(data),
    })
    if (!response.ok) {
      throw new Error(`Failed to update module: ${response.statusText}`)
    }
    return response.json()
  }

  async deleteModule(id: string): Promise<{ success: boolean; message: string }> {
    const response = await fetch(`${this.baseUrl}/${id}`, {
      method: 'DELETE',
    })
    if (!response.ok) {
      throw new Error(`Failed to delete module: ${response.statusText}`)
    }
    return response.json()
  }

  async toggleModuleStatus(id: string): Promise<ModuleResponse> {
    const response = await fetch(`${this.baseUrl}/${id}/toggle-status`, {
      method: 'POST',
    })
    if (!response.ok) {
      throw new Error(`Failed to toggle module status: ${response.statusText}`)
    }
    return response.json()
  }
}

export const modulesAPI = new ModulesAPI()